### PR TITLE
Implement initializeThread/disposeThread and allow to create foreign threads

### PR DESCRIPTION
**Pull request by eregon**
* Foreign threads (threads not created by Ruby) do not register with the
  SafepointManager since we do not know if they execute Truffle code
  within a call to Context.eval() or regular non-Truffle Java code.
  If the latter, we cannot possibly stop them or ask to come to a
  guest-language safepoint.
* Each new foreign thread creates a new Ruby Thread in the root ThreadGroup.
* Add a couple tests using foreign threads.
* Do not depend on the value of the current Thread for initialization and cleanup.

### CODE REVIEW

File: src/main/java/org/truffleruby/core/thread/ThreadManager.java [hunk 1]
```diff
@@ -50,35 +50,52 @@
...
+
+    private final Map<Thread, DynamicObject> foreignThreadMap = new ConcurrentHashMap<>();
+    private final ThreadLocal<DynamicObject> currentThread = ThreadLocal.withInitial(() -> foreignThreadMap.get(Thread.currentThread()));
+
+    private final Set<DynamicObject> runningRubyThreads =
+            Collections.newSetFromMap(new ConcurrentHashMap<DynamicObject, Boolean>());
 
-    private final Set<DynamicObject> runningRubyThreads
-            = Collections.newSetFromMap(new ConcurrentHashMap<DynamicObject, Boolean>());
+    private final Set<Thread> threadsWeCreated =
```
Comment (bjfish) [inline src/main/java/org/truffleruby/core/thread/ThreadManager.java]: Is there a more descriptive name for this? E.g.  `rubyManagedThreads`, `runningRubyManagedThreads`

Comment (eregon) [reply src/main/java/org/truffleruby/core/thread/ThreadManager.java]: :+1: `rubyManagedThreads` sounds good and is consistent with the predicate method.

Comment (eregon) [reply src/main/java/org/truffleruby/core/thread/ThreadManager.java]: Done.

**Comment by chrisseaton** [comment]
Does this completely solve the issues we have with the new API and threads from foreign languages? Or is there still work to do?

**Pull request merged**